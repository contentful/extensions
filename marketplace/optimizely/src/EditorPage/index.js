/* eslint-disable jsx-a11y/no-static-element-interactions, jsx-a11y/click-events-have-key-events, jsx-a11y/anchor-is-valid */
import React, { useState, useEffect, useCallback } from 'react';
import PropTypes from 'prop-types';
import { css } from 'emotion';
import useMethods from 'use-methods';
import tokens from '@contentful/forma-36-tokens';
import { Note, Paragraph, Modal } from '@contentful/forma-36-react-components';
import StatusBar from './subcomponents/status-bar';
import ReferencesSection from './subcomponents/references-section';
import ExperimentSection from './subcomponents/experiment-section';
import VariationsSection from './subcomponents/variations-section';
import SectionSplitter from './subcomponents/section-splitter';
import { SDKContext, GlobalStateContext } from './subcomponents/all-context';
import prepareReferenceInfo, { COMBINED_LINK_VALIDATION_CONFLICT } from './reference-info';
import useInterval from '@use-it/interval';
import ConnectButton from '../ConnectButton';

const styles = {
  root: css({
    margin: tokens.spacingXl
  }),
  paragraph: css({
    marginBottom: tokens.spacingM
  }),
  link: css({
    cursor: 'pointer',
    textDecoration: 'underline'
  })
};

const methods = state => {
  return {
    setInitialData({ experiments, contentTypes, referenceInfo }) {
      state.experiments = experiments;
      state.contentTypes = contentTypes;
      state.referenceInfo = referenceInfo;
      state.loaded = true;
    },
    setError(message) {
      state.error = message;
    },
    setExperimentId(id) {
      state.experimentId = id;
      state.meta = {};
    },
    setVariations(variations) {
      state.variations = variations;
    },
    setEntry(id, entry) {
      state.entries[id] = entry;
    },
    setMeta(meta) {
      state.meta = meta;
    },
    setExperimentResults(id, results) {
      state.experimentsResults[id] = results;
    },
    updateExperiment(id, experiment) {
      const index = state.experiments.findIndex(
        experiment => experiment.id.toString() === id.toString()
      );
      if (index !== -1) {
        state.experiments[index] = experiment;
      }
    }
  };
};

const getInitialValue = sdk => ({
  loaded: false,
  error: false,
  experiments: [],
  contentTypes: [],
  meta: sdk.entry.fields.meta.getValue() || {},
  variations: sdk.entry.fields.variations.getValue() || [],
  experimentId: sdk.entry.fields.experimentId.getValue(),
  entries: {},
  experimentsResults: {}
});

const fetchInitialData = async (sdk, client) => {
  const { space, ids, locales } = sdk;

  const [contentTypesRes, entriesRes, experiments] = await Promise.all([
    space.getContentTypes(),
    space.getEntries({ links_to_entry: ids.entry, skip: 0, limit: 1000 }),
    client.getExperiments()
  ]);

  return {
    experiments,
    contentTypes: contentTypesRes.items,
    referenceInfo: prepareReferenceInfo({
      contentTypes: contentTypesRes.items,
      entries: entriesRes.items,
      variationContainerId: ids.entry,
      variationContainerContentTypeId: ids.contentType,
      defaultLocale: locales.default
    })
  };
};

function isCloseToExpiration(expires) {
  const _10minutes = 600000;
  return parseInt(expires, 10) - Date.now() <= _10minutes;
}

export default function EditorPage(props) {
  const globalState = useMethods(methods, getInitialValue(props.sdk));
  const [state, actions] = globalState;
  const [showAuth, setShowAuth] = useState(isCloseToExpiration(props.expires));

  const experiment = state.experiments.find(
    experiment => experiment.id.toString() === state.experimentId
  );

  /**
   * Fetch initial portion of data required to render initial state
   */
  useEffect(() => {
    fetchInitialData(props.sdk, props.client)
      .then(data => {
        actions.setInitialData(data);
        return data;
      })
      .catch(() => {
        actions.setError('Unable to load initial data');
      });
  }, [actions, props.client, props.sdk]);

  /**
   * Pulling current experiment every 5s to get new status and variations
   */
  useInterval(() => {
    if (state.experimentId) {
      props.client
        .getExperiment(state.experimentId)
        .then(experiment => {
          actions.updateExperiment(state.experimentId, experiment);
          return experiment;
        })
        .catch(() => {});
    }
  }, 5000);

  /*
   * Poll to see if we need to show the reauth flow preemptively
   */
  useInterval(() => {
    setShowAuth(isCloseToExpiration(props.expires));
  }, 5000);

  /**
   * Subscribe for changes in entry
   */
  useEffect(() => {
    const unsubsribeExperimentChange = props.sdk.entry.fields.experimentId.onValueChanged(data => {
      actions.setExperimentId(data);
    });
    const unsubsribeVariationsChange = props.sdk.entry.fields.variations.onValueChanged(data => {
      actions.setVariations(data || []);
    });
    const unsubscribeMetaChange = props.sdk.entry.fields.meta.onValueChanged(data => {
      actions.setMeta(data || {});
    });
    return () => {
      unsubsribeExperimentChange();
      unsubsribeVariationsChange();
      unsubscribeMetaChange();
    };
  }, [
    actions,
    props.sdk.entry.fields.experimentId,
    props.sdk.entry.fields.meta,
    props.sdk.entry.fields.variations
  ]);

  /**
   * Update title every time experiment is changed
   */
  useEffect(() => {
    if (state.loaded) {
      const title = experiment ? `[Optimizely] ${experiment.name}` : '';
      props.sdk.entry.fields.experimentTitle.setValue(title);
    }
  }, [experiment, props.sdk.entry.fields.experimentTitle, state.loaded]);

  /**
   * Fetch experiment results every time experiment is changed
   */
  useEffect(() => {
    if (state.loaded && experiment) {
      props.client
        .getExperimentResults(experiment.id)
        .then(results => {
          actions.setExperimentResults(experiment.id, results);
          return results;
        })
        .catch(() => {});
    }
  }, [actions, experiment, props.client, state.loaded]);

  const getExperimentResults = experiment => {
    if (!experiment) {
      return undefined;
    }
    return {
      url: props.client.getResultsUrl(experiment.campaign_id, experiment.id),
      results: state.experimentsResults[experiment.id]
    };
  };

  /**
   * Handlers
   */

  const onChangeExperiment = useCallback(value => {
    props.sdk.entry.fields.meta.setValue({});
    props.sdk.entry.fields.experimentId.setValue(value.experimentId);
    props.sdk.entry.fields.experimentKey.setValue(value.experimentKey);
  });

  const onLinkVariation = useCallback(async variation => {
    const data = await props.sdk.dialogs.selectSingleEntry({
      locale: props.sdk.locales.default,
      contentTypes: state.referenceInfo.linkContentTypes
    });

    if (!data) {
      return;
    }

    const values = props.sdk.entry.fields.variations.getValue() || [];
    const meta = props.sdk.entry.fields.meta.getValue() || {};
    props.sdk.entry.fields.meta.setValue({
      ...meta,
      [variation.key]: data.sys.id
    });
    props.sdk.entry.fields.variations.setValue([
      ...values,
      {
        sys: {
          type: 'Link',
          id: data.sys.id,
          linkType: 'Entry'
        }
      }
    ]);
  });

  const onOpenEntry = useCallback(entryId => {
    props.sdk.navigator.openEntry(entryId, { slideIn: true });
  });

  const onCreateVariation = useCallback(async (variation, contentTypeId) => {
    const data = await props.sdk.navigator.openNewEntry(contentTypeId, {
      slideIn: true
    });

    if (!data) {
      return;
    }

    const values = props.sdk.entry.fields.variations.getValue() || [];
    const meta = props.sdk.entry.fields.meta.getValue() || {};

    props.sdk.entry.fields.meta.setValue({
      ...meta,
      [variation.key]: data.entity.sys.id
    });
    props.sdk.entry.fields.variations.setValue([
      ...values,
      {
        sys: {
          type: 'Link',
          id: data.entity.sys.id,
          linkType: 'Entry'
        }
      }
    ]);
  });

  const onRemoveVariation = useCallback((entryId, variation) => {
    const values = props.sdk.entry.fields.variations.getValue() || [];
    const meta = props.sdk.entry.fields.meta.getValue() || {};
    if (variation) {
      delete meta[variation.key];
    }
    props.sdk.entry.fields.meta.setValue(meta);
    props.sdk.entry.fields.variations.setValue(values.filter(item => item.sys.id !== entryId));
  });

  const onClearVariations = useCallback(() => {
    props.sdk.entry.fields.meta.setValue({});
    props.sdk.entry.fields.variations.setValue([]);
  });

  const { combinedLinkValidationType } = state.referenceInfo || {};
  if (combinedLinkValidationType === COMBINED_LINK_VALIDATION_CONFLICT) {
    return (
      <Note noteType="negative" title="Conflict">
        Validations of reference fields in incoming references yield conflicting references for the
        Variation Container. Loosen validations or change incoming references so there is at least
        one shared Content Type validation.
      </Note>
    );
  }

  return (
    <SDKContext.Provider value={props.sdk}>
      <GlobalStateContext.Provider value={globalState}>
        <Modal title="Connect with Optimizely" isShown={!props.client}>
          <Paragraph className={styles.paragraph} testId="reconnect-optimizely">
            Your Optimizely session has expired. Reconnect to continue editing.
          </Paragraph>
          <ConnectButton openAuth={props.openAuth} />
        </Modal>
        <div className={styles.root} data-test-id="editor-page">
          <StatusBar
            loaded={state.loaded}
            experiment={experiment}
            variations={state.variations}
            entries={state.entries}
          />
          <SectionSplitter />
          {showAuth && (
            <Note noteType="warning" className={styles.paragraph}>
              Your Optimizely session will expire soon. Click here to{' '}
              <a onClick={props.openAuth} className={styles.link} data-test-id="preemptive-connect">
                connect with Optimizely.
              </a>
            </Note>
          )}
          <ReferencesSection
            loaded={state.loaded}
            references={state.loaded ? state.referenceInfo.references : []}
            sdk={props.sdk}
          />
          <SectionSplitter />
          <ExperimentSection
            loaded={state.loaded}
            disabled={experiment && state.variations.length > 0}
            experiments={state.experiments}
            experiment={experiment}
            onChangeExperiment={onChangeExperiment}
            onClearVariations={onClearVariations}
          />
          <SectionSplitter />
          <VariationsSection
            loaded={state.loaded}
            contentTypes={state.contentTypes}
            experiment={experiment}
            experimentResults={getExperimentResults(experiment)}
            meta={state.meta}
            variations={state.variations}
            onCreateVariation={onCreateVariation}
            onLinkVariation={onLinkVariation}
            onOpenEntry={onOpenEntry}
            onRemoveVariation={onRemoveVariation}
          />
        </div>
      </GlobalStateContext.Provider>
    </SDKContext.Provider>
  );
}

EditorPage.propTypes = {
  openAuth: PropTypes.func.isRequired,
  client: PropTypes.any.isRequired,
  expires: PropTypes.string.isRequired,
  sdk: PropTypes.shape({
    space: PropTypes.object.isRequired,
    ids: PropTypes.object.isRequired,
    locales: PropTypes.object.isRequired,
    navigator: PropTypes.shape({
      openEntry: PropTypes.func.isRequired,
      openNewEntry: PropTypes.func.isRequired
    }).isRequired,
    dialogs: PropTypes.shape({
      selectSingleEntry: PropTypes.func.isRequired
    }).isRequired,
    entry: PropTypes.shape({
      fields: PropTypes.shape({
        experimentId: PropTypes.object.isRequired,
        experimentKey: PropTypes.object.isRequired,
        variations: PropTypes.object.isRequired,
        meta: PropTypes.object.isRequired,
        experimentTitle: PropTypes.object.isRequired
      }).isRequired,
      getSys: PropTypes.func.isRequired
    }).isRequired,
    parameters: PropTypes.shape({
      installation: PropTypes.shape({
        optimizelyProjectId: PropTypes.string.isRequired
      }).isRequired
    }).isRequired
  }).isRequired
};
