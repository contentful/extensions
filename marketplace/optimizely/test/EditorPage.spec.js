import React from 'react';
import { cleanup, render, wait, configure } from '@testing-library/react';

import mockProps from './mockProps';
import mockVariantData from './mockData/mockVariantData';

import EditorPage from '../src/EditorPage';

configure({ testIdAttribute: 'data-test-id' });

describe('EditorPage', () => {
  afterEach(cleanup);

  it('should show the reauth modal when no client is available', () => {
    const { getByTestId } = render(<EditorPage sdk={mockProps.sdk} />);

    expect(getByTestId('reconnect-optimizely')).toMatchSnapshot();
  });

  it('should show the preemtive reconnect warning box', () => {
    const expires = (Date.now() + 50000).toString();
    const { getByTestId } = render(
      <EditorPage sdk={mockProps.sdk} client={() => {}} expires={expires} />
    );

    expect(getByTestId('preemptive-connect')).toMatchSnapshot();
  });

  it('should show the experiment data when loaded', async () => {
    const space = {
      getContentTypes: () => Promise.resolve(mockVariantData.contentTypes),
      getEntries: () => Promise.resolve(mockVariantData.entries)
    };

    let valueChange = null;

    const sdk = { ...mockProps.sdk, space };

    sdk.entry.fields.experimentTitle.setValue = jest.fn();
    sdk.entry.fields.experimentId.onValueChanged = fn => {
      valueChange = fn;
      return () => {};
    };

    const client = {
      getExperiments: () => Promise.resolve(mockVariantData.experiments),
      getResultsUrl: (campaignUrl, experimentId) => {
        return `https://app.optimizely.com/v2/projects/123/results/${campaignUrl}/experiments/${experimentId}`;
      },
      getExperimentResults: () => Promise.resolve(mockVariantData.results)
    };

    const expires = (Date.now + 10000000).toString();
    const { container } = render(
      <EditorPage sdk={sdk} expires={expires} client={client} openAuth={() => {}} />
    );

    await wait();
    valueChange('15049730511');

    await wait();

    expect(container).toMatchSnapshot();
  });
});
