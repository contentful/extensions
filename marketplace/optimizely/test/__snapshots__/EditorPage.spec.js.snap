// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorPage should show the experiment data when loaded 1`] = `
<div>
  <div
    class="css-175t7kh"
    data-test-id="editor-page"
  >
    <div
      class="css-3p38p2"
    >
      <div
        class="css-xieh1a"
      >
        <svg
          class="Icon__Icon___38Epv css-uoder Icon__Icon--small___1yGZK Icon__Icon--positive___1V4nP"
          data-test-id="cf-ui-icon"
          height="18"
          viewBox="0 0 24 24"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
        <span>
          Select experiment
        </span>
      </div>
      <svg
        class="Icon__Icon___38Epv css-13af3fo Icon__Icon--small___1yGZK Icon__Icon--muted___3egnD"
        data-test-id="cf-ui-icon"
        height="18"
        viewBox="0 0 24 24"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
      <div
        class="css-xieh1a"
      >
        <svg
          class="Icon__Icon___38Epv css-uoder Icon__Icon--small___1yGZK Icon__Icon--muted___3egnD"
          data-test-id="cf-ui-icon"
          height="18"
          viewBox="0 0 24 24"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
          />
        </svg>
        <span>
          Add content
        </span>
      </div>
      <svg
        class="Icon__Icon___38Epv css-13af3fo Icon__Icon--small___1yGZK Icon__Icon--muted___3egnD"
        data-test-id="cf-ui-icon"
        height="18"
        viewBox="0 0 24 24"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
      <div
        class="css-xieh1a"
      >
        <svg
          class="Icon__Icon___38Epv css-uoder Icon__Icon--small___1yGZK Icon__Icon--muted___3egnD"
          data-test-id="cf-ui-icon"
          height="18"
          viewBox="0 0 24 24"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
          />
        </svg>
        <span>
          Publish variations
        </span>
      </div>
      <svg
        class="Icon__Icon___38Epv css-13af3fo Icon__Icon--small___1yGZK Icon__Icon--muted___3egnD"
        data-test-id="cf-ui-icon"
        height="18"
        viewBox="0 0 24 24"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
      <div
        class="css-xieh1a"
      >
        <svg
          class="Icon__Icon___38Epv css-uoder Icon__Icon--small___1yGZK Icon__Icon--positive___1V4nP"
          data-test-id="cf-ui-icon"
          height="18"
          viewBox="0 0 24 24"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
        <span>
          Start experiment
        </span>
      </div>
    </div>
    <hr
      class="css-14380qe"
    />
    <h2
      class="Heading__Heading___83D3K css-1fobf8d"
      data-test-id="cf-ui-heading"
    >
      Referenced in:
    </h2>
    <div
      class="css-1y5g446"
    >
      <p
        class="Paragraph__Paragraph___2aO-9 css-acwcvw"
        data-test-id="cf-ui-paragraph"
      >
        No other entries link to this entry.
      </p>
    </div>
    <hr
      class="css-14380qe"
    />
    <h2
      class="Heading__Heading___83D3K css-1fobf8d"
      data-test-id="cf-ui-heading"
    >
      Experiment:
    </h2>
    <div
      class="SelectField__SelectField___kbQlf"
      data-test-id="cf-ui-select-field"
    >
      <div
        class="SelectField__SelectField__label-wrapper___3jGwo"
      >
        <label
          class="FormLabel__FormLabel___3d6zQ"
          data-test-id="cf-ui-form-label"
          for="experiment"
        >
          Optimizely experiment
          <span
            class="FormLabel__FormLabel__required-text___3mvdm"
          >
            (
            required
            )
          </span>
        </label>
      </div>
      <div
        class="Select__Select__wrapper___2pBIz Select__Select--large___1PgJW"
      >
        <select
          aria-label="experiment"
          class="Select__Select___31Z46 a11y__focus-border--default___60AXp"
          data-test-id="cf-ui-select"
          id="experiment"
          name="experiment"
          required=""
        >
          <option
            data-test-id="cf-ui-select-option"
            value="-1"
          >
            Select Optimizely experiment
          </option>
          <option
            data-test-id="cf-ui-select-option"
            value="15049730511"
          >
            landing-page-hero (running)
          </option>
        </select>
        <svg
          class="Icon__Icon___38Epv Select__Select__icon___boRAZ Icon__Icon--small___1yGZK Icon__Icon--muted___3egnD"
          data-test-id="cf-ui-icon"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
        </svg>
      </div>
    </div>
    
    <hr
      class="css-14380qe"
    />
    <h2
      class="Heading__Heading___83D3K"
      data-test-id="cf-ui-heading"
    >
      Variations:
    </h2>
    <div
      class="css-1pfvl5o"
    >
      <p
        class="Paragraph__Paragraph___2aO-9"
        data-test-id="cf-ui-paragraph"
      >
        Content created in this experiment is only available for this experiment.
      </p>
      <div
        class="css-i51og3"
      >
        <h2
          class="Subheading__Subheading___2mA9j css-f6z6nw"
          data-test-id="cf-ui-subheading"
        >
          variation_1
           
          <small>
            (
            50
            % of traffic)
          </small>
        </h2>
        <div
          class="css-15g2oxy"
        >
          <div
            class="css-cet0rr"
          >
            <div
              class="Dropdown__Dropdown___1qyn8"
              data-test-id="cf-ui-dropdown"
              position="bottom-left"
            >
              <button
                class="TextLink__TextLink___1biUr a11y__focus-outline--default___2hwb1 TextLink__TextLink--primary___2Vc9F"
                data-test-id="cf-ui-text-link"
                type="button"
              >
                <span
                  class="TabFocusTrap__TabFocusTrap___39Vty"
                  tabindex="-1"
                >
                  <div
                    class="TextLink__TextLink__icon-wrapper___25_tI"
                  >
                    <svg
                      class="Icon__Icon___38Epv TextLink__TextLink__icon___3ggiB Icon__Icon--small___1yGZK undefined"
                      data-test-id="cf-ui-icon"
                      height="18"
                      viewBox="0 0 24 24"
                      width="18"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                      />
                      <path
                        d="M0 0h24v24H0z"
                        fill="none"
                      />
                    </svg>
                  </div>
                  Create entry and link
                </span>
              </button>
            </div>
          </div>
          <div
            class="css-cet0rr"
          >
            <button
              class="TextLink__TextLink___1biUr a11y__focus-outline--default___2hwb1 TextLink__TextLink--primary___2Vc9F"
              data-test-id="cf-ui-text-link"
              type="button"
            >
              <span
                class="TabFocusTrap__TabFocusTrap___39Vty"
                tabindex="-1"
              >
                <div
                  class="TextLink__TextLink__icon-wrapper___25_tI"
                >
                  <svg
                    class="Icon__Icon___38Epv TextLink__TextLink__icon___3ggiB Icon__Icon--small___1yGZK undefined"
                    data-test-id="cf-ui-icon"
                    height="18"
                    viewBox="0 0 24 24"
                    width="18"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7c-2.76 0-5 2.24-5 5s2.24 5 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1zM8 13h8v-2H8v2zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4c2.76 0 5-2.24 5-5s-2.24-5-5-5z"
                    />
                  </svg>
                </div>
                Link an existing entry
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        class="css-i51og3"
      >
        <h2
          class="Subheading__Subheading___2mA9j css-f6z6nw"
          data-test-id="cf-ui-subheading"
        >
          variation_2
           
          <small>
            (
            50
            % of traffic)
          </small>
        </h2>
        <div
          class="css-15g2oxy"
        >
          <div
            class="css-cet0rr"
          >
            <div
              class="Dropdown__Dropdown___1qyn8"
              data-test-id="cf-ui-dropdown"
              position="bottom-left"
            >
              <button
                class="TextLink__TextLink___1biUr a11y__focus-outline--default___2hwb1 TextLink__TextLink--primary___2Vc9F"
                data-test-id="cf-ui-text-link"
                type="button"
              >
                <span
                  class="TabFocusTrap__TabFocusTrap___39Vty"
                  tabindex="-1"
                >
                  <div
                    class="TextLink__TextLink__icon-wrapper___25_tI"
                  >
                    <svg
                      class="Icon__Icon___38Epv TextLink__TextLink__icon___3ggiB Icon__Icon--small___1yGZK undefined"
                      data-test-id="cf-ui-icon"
                      height="18"
                      viewBox="0 0 24 24"
                      width="18"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                      />
                      <path
                        d="M0 0h24v24H0z"
                        fill="none"
                      />
                    </svg>
                  </div>
                  Create entry and link
                </span>
              </button>
            </div>
          </div>
          <div
            class="css-cet0rr"
          >
            <button
              class="TextLink__TextLink___1biUr a11y__focus-outline--default___2hwb1 TextLink__TextLink--primary___2Vc9F"
              data-test-id="cf-ui-text-link"
              type="button"
            >
              <span
                class="TabFocusTrap__TabFocusTrap___39Vty"
                tabindex="-1"
              >
                <div
                  class="TextLink__TextLink__icon-wrapper___25_tI"
                >
                  <svg
                    class="Icon__Icon___38Epv TextLink__TextLink__icon___3ggiB Icon__Icon--small___1yGZK undefined"
                    data-test-id="cf-ui-icon"
                    height="18"
                    viewBox="0 0 24 24"
                    width="18"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7c-2.76 0-5 2.24-5 5s2.24 5 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1zM8 13h8v-2H8v2zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4c2.76 0 5-2.24 5-5s-2.24-5-5-5z"
                    />
                  </svg>
                </div>
                Link an existing entry
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`EditorPage should show the preemtive reconnect warning box 1`] = `
<a
  class="css-1lglquy"
  data-test-id="preemptive-connect"
>
  connect with Optimizely.
</a>
`;

exports[`EditorPage should show the reauth modal when no client is available 1`] = `
<p
  class="Paragraph__Paragraph___2aO-9 css-acwcvw"
  data-test-id="reconnect-optimizely"
>
  Your Optimizely session has expired. Reconnect to continue editing.
</p>
`;
